contracts/protocol/clients/DRFeeMutualizer.sol

```diff
@@ -0,0 +1,559 @@
+        if (poolBalances[_tokenAddress] < _amount) revert InsufficientPoolBalance();
+
+        unchecked {
+            poolBalances[_tokenAddress] -= _amount;
+        }
+
+        transferFundsOut(_tokenAddress, _to, _amount);
+
+        emit FundsWithdrawn(_to, _tokenAddress, _amount);
+    }
+
+    /**
+     * @notice Gets pool balance for a token
+     * @param _tokenAddress The token address (address(0) for native currency)
+     * @return balance The pool balance
+     */
+    function getPoolBalance(address _tokenAddress) external view returns (uint256 balance) {
+        return poolBalances[_tokenAddress];
+    }
+
+    // ============= Agreement Management =============
+
+    /**
+     * @notice Creates a new agreement between seller and dispute resolver for a specific exchange token
+     * @param _sellerId The seller ID
+     * @param _tokenAddress The exchange token address for the agreement (address(0) for native currency)
+     * @param _disputeResolverId The dispute resolver ID (0 for "any dispute resolver" i.e. universal agreement)
+     * @param _maxAmountPerTx The maximum mutualized amount per transaction
+     * @param _maxAmountTotal The maximum total mutualized amount
+     * @param _timePeriod The time period for the agreement (in seconds)
+     * @param _premium The premium amount to be paid by seller
+     * @param _refundOnCancel Whether premium is refunded on cancellation
+     * @return agreementId The ID of the created agreement
+     * @dev Only callable by the contract owner. Prevents duplicate active agreements for the same seller, token and dispute resolver.
+     *      Universal agreements can be created by setting disputeResolverId=0 (covers all dispute resolvers for that token).
+     *
+     * Reverts if:
+     * - Caller is not owner
+     * - sellerId is 0
+     * - maxAmountPerTx is 0
+     * - maxAmountTotal < maxAmountPerTx
+     * - timePeriod is 0
+     * - Active agreement exists for same seller, exchange token and dispute resolver
+     */
+    function newAgreement(
+        uint256 _sellerId,
+        address _tokenAddress,
+        uint256 _disputeResolverId,
+        uint256 _maxAmountPerTx,
+        uint256 _maxAmountTotal,
+        uint256 _timePeriod,
+        uint256 _premium,
+        bool _refundOnCancel
+    ) external onlyOwner returns (uint256 agreementId) {
+        if (_sellerId == 0) revert InvalidSellerId();
+        if (_maxAmountPerTx == 0) revert MaxAmountPerTxMustBeGreaterThanZero();
+        if (_maxAmountTotal < _maxAmountPerTx) revert MaxTotalMustBeGreaterThanOrEqualToMaxPerTx();
+        if (_timePeriod == 0) revert TimePeriodMustBeGreaterThanZero();
+
+        uint256 existingAgreementId = sellerToTokenToDisputeResolverToAgreement[_sellerId][_tokenAddress][
+            _disputeResolverId
+        ];
+        if (existingAgreementId != 0) {
+            Agreement storage existingAgreement = agreements[existingAgreementId];
+            if (
```

Comment:
If there is an already existing agreement for the same sellerId/tokenAddress/disputeResolferId, but not started yet (existingAgreement.startTime == 0), the method doesn't revert, creates a new agreement and replaces the agreementId in the sellerToTokenToDisputeResolverToAgreement map.
Not sure this behaviour is the one expected.
We could update the existing (not started) agreement with the new params, instead of created a new one.

Reply:
Maybe better => we can create a new agreement, after automatically voiding the previous one, so that we act its replacement with the new one.

Reply:
That's a good observation. Just my 2c:
- we don't want to update the existing agreement, since that would open up front-running attacks
- the real problem occurs if `newAgreement` and `payPremium` are racing. If `newAgreement` is processed 1s, the seller would pay for the old agreement which is not referenced in `sellerToTokenToDisputeResolverToAgreement` anymore.
- your 2nd suggestion seems ok. If `newAgreement` and `payPremium` are racing, whichever will be processed 2nd will revert.